plugins/auth-backend/src/providers/bitbucket/provider.ts

```diff
@@ -0,0 +1,313 @@
+      providerInfo: {
+        idToken: result.params.id_token,
+        accessToken: result.accessToken,
+        scope: result.params.scope,
+        expiresInSeconds: result.params.expires_in,
+      },
+      profile,
+    };
+
+    if (this.signInResolver) {
+      response.backstageIdentity = await this.signInResolver(
+        {
+          result,
+          profile,
+        },
+        {
+          tokenIssuer: this.tokenIssuer,
+          catalogIdentityClient: this.catalogIdentityClient,
+          logger: this.logger,
+        },
+      );
+    }
+
+    return response;
+  }
+}
+
+export const bitbucketEmailSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { profile } = info;
+
+  if (!profile.email) {
+    throw new Error('Bitbucket profile contained no email');
+  }
+
+  const entity = await ctx.catalogIdentityClient.findUser({
+    annotations: {
+      'bitbucket.org/email': profile.email,
+    },
+  });
+
+  const claims = getEntityClaims(entity);
+  const token = await ctx.tokenIssuer.issueToken({ claims });
+
+  return { id: entity.metadata.name, entity, token };
+};
+
+const bitbucketDefaultSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { fullProfile } = info.result;
+
+  const userId = fullProfile.username || fullProfile.id;
+
+  const token = await ctx.tokenIssuer.issueToken({
+    claims: { sub: userId, ent: [`user:default/${userId}`] },
+  });
+
+  return { id: userId, token };
+};
+
+export type BitbucketProviderOptions = {
+  /**
+   * The profile transformation function used to verify and convert the auth response
+   * into the profile that will be presented to the user.
+   */
+  authHandler?: AuthHandler<OAuthResult>;
+
+  /**
+   * Configure sign-in for this provider, without it the provider can not be used to sign users in.
+   */
+  signIn: {
+    /**
+     * Maps an auth result to a Backstage identity for the user.
+     */
+    resolver: SignInResolver<OAuthResult>;
+  };
+};
+
+export const createBitbucketProvider = (
+  options?: BitbucketProviderOptions,
+): AuthProviderFactory => {
+  return ({
+    providerId,
+    globalConfig,
+    config,
+    tokenIssuer,
+    catalogApi,
+    logger,
+  }) =>
+    OAuthEnvironmentHandler.mapConfig(config, envConfig => {
+      const clientId = envConfig.getString('clientId');
+      const clientSecret = envConfig.getString('clientSecret');
+      const callbackUrl = `${globalConfig.baseUrl}/${providerId}/handler/frame`;
+
+      const catalogIdentityClient = new CatalogIdentityClient({
+        catalogApi,
+        tokenIssuer,
+      });
+
+      const authHandler: AuthHandler<OAuthResult> = options?.authHandler
+        ? options.authHandler
+        : async ({ fullProfile, params }) => ({
+            profile: makeProfileInfo(fullProfile, params.id_token),
+          });
+
+      const signInResolverFn =
+        options?.signIn?.resolver ?? bitbucketDefaultSignInResolver;
```

Comment:
I think this fell between the cracks a bit too. Let's completely remove the `bitbucketDefaultSignInResolver` and not have any default resolver

Reply:
I've done the types.d.ts, before I push I'll add this as well. Sorry I don't see how removing the default resolver works. Above quoted line is the precise mechanism allowing user to plug in the Email-resolver aka. `'bitbucket.org/email': profile.email` annotation. 
The default at the moment uses : `fullProfile.username || fullProfile.id;` which I think would be preferable to most? 
By removing do you mean restructuring so that the username/id logic still remains the default?

Reply:
Removing it as the default would be to just change the line to `const signInResolverFn = options?.signIn?.resolver`

Part of the change we're making with the sign-in resolvers is to not have the auth providers act as sign-in providers by default. We've avoided that breaking change for now as we migrate existing providers by just having a default resolver that implements whatever the existing logic is. The default resolvers aren't supposed to stick around though, and once we migrated all auth providers we remove all defaults of them in a breaking change.

Good point about this default resolver being useful though. I think it could make sense to keep, but let's call it something that describes how it works, something like `bitbucketUserIdSignInResolver`? We'd also need to add the call out to the catalog and `getEntityClaims` to make sure it creates a proper identity that exists.

Reply:
@Rugvip cool I'll make bitbucketUserIdSignInResolver with call to catalog and getEntityClaims, do you think it's good idea to create a user in catalog if missing? Like new joiners first logging in first time? Or rely on other ingestion processor? Though I think the processor polling is being moved away from due to eating up rate limits. Cheers

Reply:
That's not possible at the moment unfortunately. The users need to already be present in the catalog, ingested via some other means

Reply:
3 resolvers are now added, and default removed
  bitbucketUsernameSignInResolver,
  bitbucketUuidSignInResolver,
  bitbucketDisplayNameSignInResolver,
annotations:
    bitbucket.org/uuid: "{80dc88eXXXXXXcc5265}"
OR
annotations:
    bitbucket.org/displayName: Filip Swiatczak
OR
annotations:
    bitbucket.org/username: filip_swiatczak
    
So user decides which one to use and creates corresponding annotations! I will add more info in the main description